vaadin-flow-components-shared-parent/vaadin-flow-components-base/src/main/java/com/vaadin/flow/component/shared/internal/DisableOnClickController.java

```diff
@@ -0,0 +1,135 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+package com.vaadin.flow.component.shared.internal;
+
+import java.io.Serializable;
+import java.util.Objects;
+
+import com.vaadin.flow.component.ClickNotifier;
+import com.vaadin.flow.component.Component;
+import com.vaadin.flow.component.HasEnabled;
+import com.vaadin.flow.component.dependency.JsModule;
+import com.vaadin.flow.component.page.PendingJavaScriptResult;
+
+/**
+ * An internal controller for handling disabling a component when it is clicked.
+ * Not intended to be used publicly.
+ * <p>
+ * When {@link #setDisableOnClick(boolean)} is enabled, the component will be
+ * immediately disabled upon clicking, both on the client-side and server-side,
+ * to prevent multiple clicks or submissions while the server processes the
+ * event.
+ * <p>
+ * This controller requires that the component implements both
+ * {@link HasEnabled} and {@link ClickNotifier}.
+ *
+ * @param <C>
+ *            Type of the component that uses this controller.
+ */
+@JsModule("./disableOnClickFunctions.js")
+public class DisableOnClickController<C extends Component & HasEnabled & ClickNotifier<C>>
+        implements Serializable {
+
+    private final C component;
+    private boolean disableOnClick = false;
+    private PendingJavaScriptResult initDisableOnClick;
+
+    /**
+     * Creates a new controller for the given component.
+     * <p>
+     * The component must implement {@link HasEnabled} and
+     * {@link ClickNotifier}.
+     *
+     * @param component
+     *            the component to control, not {@code null}
+     */
+    public DisableOnClickController(C component) {
+        this.component = Objects.requireNonNull(component);
+        component.addClickListener(itemClickEvent -> {
+            if (isDisableOnClick()) {
+                component.setEnabled(false);
+            }
+        });
+        component.addAttachListener(aAttachEvent -> {
+            if (isDisableOnClick()) {
+                initDisableOnClick();
+            }
+        });
+    }
+
+    /**
+     * Sets whether the component should be disabled when clicked.
+     * <p>
+     * When set to {@code true}, the component will be immediately disabled on
+     * the client-side when clicked, preventing further clicks until re-enabled
+     * from the server-side.
+     *
+     * @param disableOnClick
+     *            whether the component should be disabled when clicked
+     */
+    public void setDisableOnClick(boolean disableOnClick) {
+        this.disableOnClick = disableOnClick;
+        if (disableOnClick) {
+            component.getElement().setAttribute("disableOnClick", "true");
+            initDisableOnClick();
```

Comment:
Would it be enough to run `initDisableOnClick()` only in the attach listener?

Reply:
We also need one to run for the case where the component is already attached, and set `disableOnClick`. Reafactored the logic to make it clearer.